File: src/hashtable/TwoSum.java
package hashtable;

import java.util.HashMap;
import java.util.Map;

public class TwoSum {
    public int[] twoSum(int[] nums, int target) {
        int[] result = new int[2];
        Map<Integer, Integer> dictionary = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int remain = target - nums[i];
            if (dictionary.containsKey(remain)) {
                result[0] = i;
                result[1] = dictionary.get(remain);
                return result;
            }
            dictionary.put(nums[i], i);
        }
        return result;
    }

    public static void main(String[] args) {
        int[] nums = {3, 2, 4};
        new TwoSum().twoSum(nums, 6);
    }
}

File: rust/src/hash_table/two_sum.rs
use super::Solution;
use std::collections::HashMap;
impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map = HashMap::new();
        for (index, val) in nums.iter().enumerate() {
            let remain = target - val;
            if let Some(&item) = map.get(&remain) {
                return vec![index as i32, item];
            }
            map.insert(*val, index as i32);
        }
        vec![]
    }
}
#[cfg(test)]
mod tests {
    use super::Solution;
    #[test]
    fn test_hash_table() {
        Solution::two_sum(vec![2, 7, 11, 15], 9);
    }
}
